DllImportGenerator/designs/StructMarshalling.md

```diff
@@ -0,0 +1,127 @@
+
+I propose an opt-in design where the owner of a struct has to explicitly opt-in to usage for interop. This enables our team to add special support as desired for various types such as `Span<T>` while also avoiding the private reflection and limited type information issues mentioned above.
+
+This design would use three attributes:
+
+```csharp
+
+public class GeneratedMarshallingAttribute : Attribute {}
+
+public class BlittableTypeAttribute : Attribute {}
+
+public class NativeStructMarshallingAttribute : Attribute
+{
+     public NativeStructMarshallingAttribute(Type nativeStructType, string managedToNativeMarshalMethod, string nativeToManagedMarshalMethod, string nativeFreeMethod) {}
+}
+
+```
+
+There are 2 usage mechanisms of these attributes.
+
+- Usage 1, Source-generated interop:
+
+The user can apply the `GeneratedMarshallingAttribute` to their structure `S`. The source generator will determine if the type is blittable. If it is blittable, the source generator will generate a partial definition and apply the `BlittableTypeAttribute` to the struct type `S`. Otherwise, it will generate a blittable representation of the struct and apply the `NativeStructMarshallingAttribute` and point it to the blittable representation, a method that marshals the managed value to the native value, a method that marshals a value of the native struct to a managed value, and a method that frees any native resources that were allocated for the native value. The blittable representation can either be generated as a separate top-level type or as a nested type on `S`.
+
+- Usage 2, Manual interop:
+
+The user may want to manually mark their types as marshalable in this system due to specific restrictions in their code base around marshaling specific types that the source generator does not account for. We could also use this internally to support custom types in source instead of in the code generator. In this scenario, the user would apply either the `BlittableTypeAttribute` or the `NativeStructMarshallingAttribute` attribute to their struct type. An analyzer would validate that the struct is blittable if the `BlittableTypeAttribute` is applied or validate that the native struct type is blittable and that the methods provided in the attribute have the correct signatures when the `NativeStructMarshallingAttribute` is applied.
+
+The P/Invoke source generator (as well as the struct source generator when nested struct types are used) would use the `BlittableTypeAttribute` and `NativeStructMarshallingAttribute` to determine how to marshal a value type parameter or field instead of looking at the fields of the struct directly.
+
+The largest problem of this design is the switch to an opt-in model for structure types. We could augment this design by providing a fallback to the built-in struct marshaling system via an attribute (either `MarshalAsAttribute` or another one) if we believe providing an explicit escape hatch is needed.
```

Comment:
This design assumes that there is one and only right way to marshal everything.

Maybe we should allow specifying the native type to use for marshalling in the PInvoke signature. Something like:

```
[DllImport(...)]
static void foo([MarshalUsing(MyNativeType)] MyType a);
```

Reply:
I think the best motivating scenario for this is to allow marshalling of non-blittable types that are owned by somebody else (I have edited my example to reflect it.)

I expect that associated functions would be found by convention on `MyNativeType` in this case.

Reply:
Hmmm. With this argument the `BlittableTypeAttribute` type becomes less interesting because if the native type is the managed type that could be interpreted as "Blittable". Right?

Reply:
Agree. `BlittableTypeAttribute` can be encoded as `NativeStructMarshallingAttribute` pointing to self.

Reply:
This may not be a good idea given the method name collisions that it can introduce (see the other comment).

Reply:
I'd be ok with encoding blittability as a `NativeStructMarshallingAttribute` pointing to self since in the blittable case we wouldn't need to use the marshalling methods (and we can't for cases where we pin).

Reply:
>This may not be a good idea given the method name collisions that it can introduce (see the other comment).

Name collisions on what? The supplied "NativeType" would be a new "transport" time simply for the moving between managed/unmanaged.

Reply:
I was referring to https://github.com/dotnet/runtimelab/pull/45#discussion_r465325937.

For example, let's say that we want mark the existing GCHandle type as blittable. 

If it is done via `[BlittableType]`, it is clear what it means. No ambiguity. This type is always passed through without any marshaling.

If it is done via `[NativeStructMarshalling(typeof(GCHandle)]`, it is less clear. For example, should `GCHandle.Free` participate in marshaling or not, assuming `Free` method is part of the convention.

Reply:
>If it is done via [NativeStructMarshalling(typeof(GCHandle)], it is less clear. For example, should GCHandle.Free participate in marshaling or not, assuming Free method is part of the convention.

Agreed that is a concern - that was part of my push back when talking about convention. However, it presupposes the need to use existing types as "Transfer types" - I like this term because it describes their purpose. For example, `GCHandle` isn't a transfer type because it already has a purpose and isn't something we would be likely to express on the unmanaged side of the boundary - we can debate that. If it is what we want to reflect on both side, then it is blittable so the collision wouldnt come into play.

Alternatively we could expand the concept mentioned above by merging it with @jkoritzinsky's original proposal.

```csharp
struct MyNameType
{
  // Pre-existing functions on the type
  public static MyNameType Alloc() { ... }
  public static long Free() { ... }

  // Support for DllImport source generator
  // N.B. Signatures are for illustrative purposes only.
  public static IntPtr MyAlloc() { ... }
  public static void MyFree(IntPtr a) { ... }
}

static void foo([MarshalUsing(MyNativeType, Alloc = nameof(MyNameType.MyAlloc), Free = nameof(MyNameType.MyFree))] TypeThatIDoNotOwn a);
```

This allows us by default to have a strongly typed explicit convention or in the cases where it doesn't work the user can say "Use this instead". I genuinely like the merging of the two ideas.